src/vs/platform/storage/common/profileStorageService.ts

```diff
@@ -0,0 +1,151 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Event } from 'vs/base/common/event';
+import { Disposable, isDisposable } from 'vs/base/common/lifecycle';
+import { IStorage, IStorageDatabase, Storage } from 'vs/base/parts/storage/common/storage';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { AbstractStorageService, IStorageService, IStorageValueChangeEvent, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { IUserDataProfile } from 'vs/platform/userDataProfile/common/userDataProfile';
+
+export interface IProfileStorageValueChanges {
+	readonly profile: IUserDataProfile;
+	readonly changes: IStorageValueChangeEvent[];
+}
+
+export interface IProfileStorageChanges {
+	readonly targetChanges: IUserDataProfile[];
+	readonly valueChanges: IProfileStorageValueChanges[];
+}
+
+export interface IStorageValue {
+	readonly value: string | undefined;
+	readonly target: StorageTarget;
+}
+
+export const IProfileStorageService = createDecorator<IProfileStorageService>('IProfileStorageService');
+export interface IProfileStorageService {
+	readonly _serviceBrand: undefined;
+
+	/**
+	 * Emitted whenever data is updated or deleted in a profile storage or target of a profile storage entry changes
+	 */
+	readonly onDidChange: Event<IProfileStorageChanges>;
+
+	/**
+	 * Return the requested profile storage data
+	 * @param profile The profile from which the data has to be read from
+	 */
+	readStorageData(profile: IUserDataProfile): Promise<Map<string, IStorageValue>>;
+
+	/**
+	 * Update the given profile storage data in the profile storage
+	 * @param profile The profile to which the data has to be written to
+	 * @param data Data that has to be updated
+	 * @param target Storage target of the data
+	 */
+	updateStorageData(profile: IUserDataProfile, data: Map<string, string | undefined | null>, target: StorageTarget): Promise<void>;
+}
+
+export abstract class AbstractProfileStorageService extends Disposable implements IProfileStorageService {
```

Comment:
Not sure how lifecycle works in this class: do we potentially write to the same DB from 2 connections (window, this), or is the connection reused? And when we dispose, do we potentially dispose a live connection of the window?